### PR TITLE
Issue 790 - Copy/paste regression when selecting cells with mouse

### LINKED ISSUE #790: Copy & paste when clicking to select cells doesn't work anymore?
Originally reported in https://community.plotly.com/t/copying-from-datatable/40571/3.

It seems like it works when you use the keyboard to navigate and select cells (shift + arrow keys) but it doesn't work for me if I use the mouse to select cells: shift + click.

I believe this worked in the past.

Here's a GIF to reproduce. Taken from https://dash.plotly.com/datatable

copy-paste

### PR DESCRIPTION
**Pull request by Marc-Andre-Rivet**
Closes #790

### CODE REVIEW

File: src/dash-table/components/ControlledTable/index.tsx [hunk 1]
```diff
@@ -429,6 +448,52 @@ export default class ControlledTable extends PureComponent<
...
+        if (
+            !active ||
+            active.column_id !== active_cell?.column_id ||
+            active.row !== active_cell?.row
+        ) {
+            return true;
+        }
+
+        return false;
+    }
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/components/ControlledTable/index.tsx:494]: I am unable to use Ctrl+C/Ctrl+V in a way that triggers copy/paste in Selenium, Chrome does not support `document.execCommand('paste')` and the ClipboardAPI requires the user to `Allow` browser permission for usage. These reasons are behind the special "test" build and also behind this long standing regression. Prior to this change, the tests worked but the real build didn't. This tries to better align the test behavior vs. the real world behavior.

File: src/dash-table/components/ControlledTable/index.tsx [hunk 2]
```diff
@@ -163,6 +163,28 @@ export default class ControlledTable extends PureComponent<
...
+                (active.column_id !== active_cell?.column_id ||
+                    active.row !== active_cell?.row)
+            ) {
+                const target = this.$el.querySelector(
+                    `td[data-dash-row="${active_cell.row}"][data-dash-column="${active_cell.column_id}"]:not(.phantom-cell)`
+                ) as HTMLElement;
+                if (target) {
+                    target.focus();
+                }
+            }
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/components/ControlledTable/index.tsx:185]: Make sure that the active cell is focused if the active element is another data cell in the table (triggered by mouse click to change the selected_cells)

File: src/dash-table/components/ControlledTable/index.tsx [hunk 3]
```diff
@@ -443,16 +508,20 @@ export default class ControlledTable extends PureComponent<
...
             /*#endif*/
             return;
         }
 
         if (e.keyCode === KEY_CODES.C && ctrlDown && !is_focused) {
             /*#if TEST_COPY_PASTE*/
-            this.onCopy(e as any);
             e.preventDefault();
+            if (!this.preventCopyPaste()) {
+                this.onCopy(e as any);
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/components/ControlledTable/index.tsx:523]: Trigger the test "copy" only if the cell with focus is the active_cell

File: tests/selenium/test_basic_copy_paste.py [hunk 4]
```diff
@@ -113,13 +114,20 @@ def test_tbcp002_sorted_copy_paste_callback(test):
...
+
+    if mouse_navigation:
+        with test.hold(Keys.SHIFT):
+            target.cell(0, 0).click()
+            target.cell(2, 0).click()
+    else:
         target.cell(0, 0).click()
-        target.cell(2, 0).click()
+        with test.hold(Keys.SHIFT):
+            test.send_keys(Keys.ARROW_DOWN + Keys.ARROW_DOWN)
```
Comment (Marc-Andre-Rivet) [inline tests/selenium/test_basic_copy_paste.py:130]: Test navigating to the cell both by selecting a region with the keyboard and the mouse

**Comment by Chris-Keo** [comment]
Hey Marc!
Any update on the copy paste issue?

**Comment by alexcjohnson** [comment]
@Chris-Keo this is a merged PR, and this work has been released. Please look through the open [issues](https://github.com/plotly/dash-table/issues) to see if the behavior you're observing has already been reported - and if not, feel free to submit a new issue and we'll take a look.

**Pull request merged**